extensions/amp-next-page/1.0/visibility-observer.js

```diff
@@ -25,36 +25,65 @@ import {
 import {RelativePositions} from '../../../src/layout-rect';
 import {Services} from '../../../src/services';
 import {devAssert} from '../../../src/log';
+import {throttle} from '../../../src/utils/rate-limit';
 
 /** @enum {number} */
 export const ViewportRelativePos = {
-  INSIDE_VIEWPORT: 0,
-  OUTSIDE_VIEWPORT: 1,
-  LEAVING_VIEWPORT: 2,
-  CONTAINS_VIEWPORT: 3,
+  INSIDE_VIEWPORT: 1,
+  OUTSIDE_VIEWPORT: 2,
+  LEAVING_VIEWPORT: 3,
+  ENTERING_VIEWPORT: 4,
+  CONTAINS_VIEWPORT: 5,
 };
 
+/** @const {number} */
+const SCROLL_DIRECTION_THRESHOLD = 20;
+
+/** @enum {number} */
+export const ScrollDirection = {UP: 1, DOWN: -1};
+
 export class VisibilityObserverEntry {
   /**
    * @param {!VisibilityObserver} observer
+   * @param {function(!ViewportRelativePos)} callback
    */
-  constructor(observer) {
+  constructor(observer, callback) {
     /** @private {!VisibilityObserver} */
     this.observer_ = observer;
     /** @private {?RelativePositions} */
     this.topSentinelPosition_ = null;
     /** @private {?RelativePositions} */
     this.bottomSentinelPosition_ = null;
-    /** @private {!ViewportRelativePos} */
-    this.relativePos_ = ViewportRelativePos.OUTSIDE_VIEWPORT;
+    /** @private {function(!ViewportRelativePos)} */
+    this.callback_ = callback;
+  }
+
+  /**
+   * @return {function(!ViewportRelativePos)}
+   */
+  get callback() {
```

Comment:
What do these getters help with?

Reply:
This is so that it's possible to call the entry's callback from the observer

Reply:
It looks like `callback` doesn't change, could you make that `@const` and provide the property directly on the constructor? Just for size. (typecheck should guard against mutating the value)

Reply:
Fixed

Reply:
What he's suggesting is that this be made into a regular property and not a getter accessor. That makes it smaller when compiled, and faster.

Reply:
Oh sounds good